dask_sql/physical/rel/logical/limit.py

```diff
@@ -50,17 +51,24 @@ def _apply_limit(self, df: dd.DataFrame, offset: int, end: int) -> dd.DataFrame:
         Unfortunately, Dask does not currently support row selection through `iloc`, so this must be done using a custom partition function.
         However, it is sometimes possible to compute this window using `head` when an `offset` is not specified.
         """
-        if not offset:
-            # We do a (hopefully) very quick check: if the first partition
-            # is already enough, we will just use this
-            first_partition_length = len(df.partitions[0])
-            if first_partition_length >= end:
-                return df.head(end, compute=False)
-
         # compute the size of each partition
         # TODO: compute `cumsum` here when dask#9067 is resolved
         partition_borders = df.map_partitions(lambda x: len(x))
 
+        # if no offset is specified we can use `head` to compute the window
```

Comment:
For comparison, the implementation I used that worked for simple queries was:
```
# if no offset is specified we can use `head` to compute the window
if not offset:
    partition_divisions = partition_borders.divisions
    try:
        ind = list(map(lambda i: i >= limit, partition_borders)).index(True)
    except ValueError:
        ind = -1
    return df.head(limit, npartitions=-1, compute=False)
```

Reply:
Thanks for sharing @sarahyurick! After further exploration of using `divisions` to keep track of partition length, I realize it might be difficult to handle certain edge cases - for example, the following input table

```python
from dask.datasets import timeseries

ddf = timeseries().set_index("x")
ddf.divisions
# (-0.9999997982529687, -0.9243231218501866, ..., 0.9017005336188877, 0.9999980053737976)
```

has divisions that don't coincide with the number of rows in each partition. Unless there's a relatively cheap way to check if a table's index is essentially an unaltered `RangeIndex`, this might push us back to using the partition lengths directly.

Reply:
Ah ok, I see what you mean